code/Hashing.cpp

```diff
@@ -0,0 +1,133 @@
+class String_Hash {
+    ll p, m, n;
+    v pre, suf, powers, inverse_powers;
+
+   public:
+    String_Hash(ll p, string &str) {
+        this->p = p;  		// p=31 for lowercase p=53 for uppercase + lowercase
+        this->m = 1e9 + 9;  // large prime
+
+        ll n = sz(str);
+        this->n = n;
+        this->pre.resize(n);
+        this->suf.resize(n);
+        this->powers.resize(n);
+        this->inverse_powers.resize(n);
+        this->powers[0] = 1;
+        ll p1 = p;
```

Comment:
We need to make this private to the loop it will be used in, and make it semantically clear that it's a copy.